airbyte-workers/src/main/java/io/airbyte/workers/temporal/PartialSuccessTemporalAttemptExecution.java

```diff
@@ -0,0 +1,221 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.workers.temporal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.functional.CheckedConsumer;
+import io.airbyte.commons.functional.CheckedSupplier;
+import io.airbyte.scheduler.models.JobRunConfig;
+import io.airbyte.workers.Worker;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/*
+ * This class allow a worker to run multiple times. In addition to the functionality in
+ * TemporalAttemptExecution it takes a predicate to determine if the output of a worker constitutes
+ * a complete success or a partial one. It also takes a function that takes in the input of the
+ * previous run of the worker and the output of the last worker in order to generate a new input for
+ * that worker.
+ */
+public class PartialSuccessTemporalAttemptExecution<INPUT, OUTPUT> implements Supplier<List<OUTPUT>> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PartialSuccessTemporalAttemptExecution.class);
+
+  private final CheckedSupplier<Worker<INPUT, OUTPUT>, Exception> workerSupplier;
+  private final Supplier<INPUT> inputSupplier;
+  private final CancellationHandler cancellationHandler;
+
+  private final Predicate<OUTPUT> shouldAttemptAgainPredicate;
+  private final BiFunction<INPUT, OUTPUT, INPUT> computeNextAttemptInputFunction;
+  private final int maxRetriesCount;
+  private final TemporalAttemptExecutionFactory<INPUT, OUTPUT> temporalAttemptExecutionFactory;
+  private final Path workspaceRoot;
+  private final JobRunConfig jobRunConfig;
+
+  public PartialSuccessTemporalAttemptExecution(Path workspaceRoot,
```

Comment:
Is the retry visible in the UI/API?

Reply:
Good question. Right now it's not. If we want to do that we need to update our tracking model (e.g. the Jobs table) to contain it. Probably worth doing, but I don't think we should block on it.

Reply:
SGTM. do you mind creating an issue describing what needs to happen to make this visible on the API?

Reply:
yes! https://github.com/airbytehq/airbyte/issues/3595